### PR TITLE
[CI/Build] Add `/opt/venv/bin` to `PATH` in `lmcache/vllm-openai` Build and Release images

### PR DESCRIPTION
**Pull request by Zerohertz**
## Summary

This PR modifies the `Dockerfiles` for both the Build and Release images of `lmcache/vllm-openai` to ensure `/opt/venv/bin` is included in the `PATH` environment variable.

## Details

- Adds `/opt/venv/bin` to the `PATH` in both the Build and Release images.
- This change ensures that executables installed in the virtual environment are accessible by default, addressing compatibility issues with the production-stack.

## Motivation

- Resolves issues where production-stack environments expect `/opt/venv/bin` executables to be available in the `PATH`.
- Improves developer and runtime experience by following standard practices for Python virtual environments.
- Reduces environment-specific setup bugs.

## Related

- Original issue: Compatibility problem with production-stack ([Slack conversation](https://lmcacheworkspace.slack.com/archives/C083AQ88Z5L/p1756812217801079))
- #1497 (replace with actual issue URL after issue creation)

## Test

```
$ docker run --rm -it --entrypoint bash lmcache/vllm-openai:build-latest 
root@c6ef938ac4e0:/workspace# which python
/opt/venv/bin/python
root@c6ef938ac4e0:/workspace# which vllm
/opt/venv/bin/vllm
```

**PLEASE READ THE CHECKLIST BELOW AND FILL IN THE DESCRIPTION ABOVE**

---

<details>

<summary><b> PR Checklist (Click to Expand) </b></summary>

<p>Thank you for your contribution to LMCache! Before submitting the pull request, please ensure the PR meets the following criteria. This helps us maintain the code quality and improve the efficiency of the review process.</p>

<h3>PR Title and Classification</h3>
<p>Please try to classify PRs for easy understanding of the type of changes. The PR title is prefixed appropriately to indicate the type of change. Please use one of the following:</p>
<ul>
    <li><code>[Bugfix]</code> for bug fixes.</li>
    <li><code>[CI/Build]</code> for build or continuous integration improvements.</li>
    <li><code>[Doc]</code> for documentation fixes and improvements.</li>
    <li><code>[Model]</code> for adding a new model or improving an existing model. Model name should appear in the title.</li>
    <li><code>[Core]</code> for changes in the core LMCache logic (e.g., <code>LMCacheEngine</code>, <code>Backend</code> etc.)</li>
    <li><code>[Misc]</code> for PRs that do not fit the above categories. Please use this sparingly.</li>
</ul>
<p><strong>Note:</strong> If the PR spans more than one category, please include all relevant prefixes.</p>

<h3>Code Quality</h3>

<p>The PR need to meet the following code quality standards:</p>

<ul>
    <li>The code need to be well-documented to ensure future contributors can easily understand the code.</li>
    <li> Please include sufficient unit tests to ensure the change is stay correct and robust. The unit and integration tests will always run and our comprehensive test will be triggered after the "full" label is tagged onto a PR.</li>
</ul>

<h3>What to Expect for the Reviews</h3>

We aim to address all PRs in a timely manner. If no one reviews your PR within 5 days, please @-mention one of KuntaiDu, ApostaC or YaoJiayi.

</details>

### CODE REVIEW

**Review by sammshen — APPROVED** [review]
LGTM!

**Comment by Zerohertz** [comment]
Is [CI failure](https://buildkite.com/lmcache/unit-tests/builds/4706#01990d09-dc94-48c3-a5d2-3d5d7e0ac677) related with this PR?

**Comment by sammshen** [comment]
I reran it and the same error occurs both times: 
```text

_____________________________________________________ test_paged_mem_leak[cpu-local_disk_remote-128-vllm] ______________________________________________________
--
  |  
  | fmt = 'vllm', chunk_size = 128, backend = 'local_disk_remote'
  | lmserver_v1_process = LMCacheServerProcess(server_url='lm://localhost:30236', server_process=<Popen: returncode: None args: ['python3', '-m', 'lmcache.v1.server', 'local...>)
  | autorelease_v1 = <function autorelease_v1.<locals>._factory at 0x7f5c0df2fc40>
  |  
  | @pytest.mark.parametrize("fmt", ["vllm"])
  | @pytest.mark.parametrize("chunk_size", [128])
  | @pytest.mark.parametrize(
  | "backend",
  | [
  | "cpu",
  | "local_disk",
  | "remote",
  | "local_disk_remote",
  | "local_cpu_disk_remote",
  | ],
  | )
  | @pytest.mark.no_shared_allocator
  | @pytest.mark.parametrize("lmserver_v1_process", ["cpu"], indirect=True)
  | def test_paged_mem_leak(fmt, chunk_size, backend, lmserver_v1_process, autorelease_v1):
  | url = None
  | if "remote" in backend:
  | url = lmserver_v1_process.server_url
  |  
  | device = "cuda"
  | num_tokens = 2000
  | kv_shape = (32, 2, chunk_size, 8, 128)
  | num_blocks = 1000
  | block_size = 16
  | dtype = torch.bfloat16
  | connector = create_gpu_connector(1024, 32)
  |  
  | tokens = generate_tokens(num_tokens, device)
  | kv_cache = generate_kv_cache_paged_list_tensors(
  | num_blocks, device, block_size, dtype
  | )
  | slot_mapping = random.sample(range(0, num_blocks * block_size), num_tokens)
  | slot_mapping = torch.tensor(slot_mapping, device=device)
  | """ initialize the engine """
  | cfg = LMCacheEngineConfig.from_legacy(
  | chunk_size=chunk_size, backend=backend, remote_url=url
  | )
  |  
  | engine = autorelease_v1(
  | LMCacheEngineBuilder.get_or_create(
  | "test",
  | cfg,
  | dumb_metadata(fmt, kv_shape),
  | connector,
  | mock_up_broadcast_fn,
  | mock_up_broadcast_object_fn,
  | )
  | )
  |  
  | engine.store(tokens, kvcaches=kv_cache, slot_mapping=slot_mapping)
  | recover_engine_states(engine)
  |  
  | expected_length = 2000
  | timeout = 30
  | """Wait until cpu store finishes"""
  | if "cpu" in backend:
  | start_time = time.time()
  | while engine.lookup(tokens, ["LocalCPUBackend"]) < expected_length:
  | if time.time() - start_time > timeout:
  | raise TimeoutError(f"Operation timed out after {timeout} seconds.")
  | time.sleep(0.01)
  | """Wait until disk store finishes"""
  | if "disk" in backend:
  | start_time = time.time()
  | while engine.lookup(tokens, ["LocalDiskBackend"]) < expected_length:
  | if time.time() - start_time > timeout:
  | raise TimeoutError(f"Operation timed out after {timeout} seconds.")
  | time.sleep(0.01)
  |  
  | if "remote" in backend:
  | start_time = time.time()
  | while engine.lookup(tokens, ["RemoteBackend"]) < expected_length:
  | if time.time() - start_time > timeout:
  | raise TimeoutError(f"Operation timed out after {timeout} seconds.")
  | time.sleep(0.01)
  | tensor_memory_allocator = (
  | engine.storage_manager.allocator_backend.memory_allocator.pin_allocator
  | )
  | if "cpu" not in backend:
  | >           assert tensor_memory_allocator.total_allocated_size == 0
  | E           assert 10485760 == 0
  | E            +  where 10485760 = <lmcache.v1.memory_management.TensorMemoryAllocator object at 0x7f5c0e08d490>.total_allocated_size
  |  
  | tests/v1/test_cache_engine.py:929: AssertionError

```

Other PRs are passing so it is this PR specifically. Though it is not immediately obvious why this PR would affect the allocator backend in unit tests, which do not use the docker image. 

I think it has something to do with the integration and unit tests being run on the same machine (will consult CI master @Shaoting-Feng  tmrw) . This is quite the interesting bug and I think it is a good thing that your PR surfaced this issue :)

File: docker/Dockerfile [hunk 1]
```diff
@@ -50,11 +50,10 @@ RUN git clone https://github.com/ai-dynamo/nixl && \
     cd nixl && \
     git checkout b1c22edd8fe10e2e5221c107ee51200fce6f09a8
 RUN cd /workspace/nixl
-RUN source /opt/venv/bin/activate
-RUN . /opt/venv/bin/activate && \
-    uv pip install meson
+ENV VIRTUAL_ENV=/opt/venv \
```
Comment (hickeyma) [inline docker/Dockerfile]: Why is `VIRTUAL_ENV` created, it is not used anywhere else?

File: docker/Dockerfile [hunk 2]
```diff
@@ -50,11 +50,10 @@ RUN git clone https://github.com/ai-dynamo/nixl && \
...
+    PATH="/opt/venv/bin:$PATH"
+RUN uv pip install meson
 RUN cd /workspace/nixl && \
-    . /opt/venv/bin/activate && \
```
Comment (hickeyma) [inline docker/Dockerfile:57]: Good catch on duplication from previously.

File: docker/Dockerfile [hunk 3]
```diff
@@ -65,15 +64,13 @@ RUN echo "/usr/local/nixl/lib/x86_64-linux-gnu" > /etc/ld.so.conf.d/nixl.conf
 RUN echo "/usr/local/nixl/lib/x86_64-linux-gnu/plugins" >> /etc/ld.so.conf.d/nixl.conf
 RUN ldconfig
 RUN cd /workspace/nixl/ && \
-    . /opt/venv/bin/activate && \
```
Comment (hickeyma) [inline docker/Dockerfile:68]: Good catch on duplication from previously.

File: docker/Dockerfile [hunk 4]
```diff
@@ -65,15 +64,13 @@ RUN echo "/usr/local/nixl/lib/x86_64-linux-gnu" > /etc/ld.so.conf.d/nixl.conf
...
     uv build --wheel --out-dir /tmp/dist && \
     uv pip install /tmp/dist/nixl-0.3.0-cp312-cp312-linux_x86_64.whl
 
 # Install runtime dependencies
 COPY ./requirements/common.txt common.txt
 COPY ./requirements/cuda.txt cuda.txt
 RUN --mount=type=cache,target=/root/.cache/pip \
-    . /opt/venv/bin/activate && \
```
Comment (hickeyma) [inline docker/Dockerfile:76]: Good catch on duplication from previously.

File: docker/Dockerfile [hunk 5]
```diff
@@ -101,7 +98,6 @@ ARG nvcc_threads=8
 ENV NVCC_THREADS=$nvcc_threads
 
 RUN --mount=type=cache,target=/root/.cache/pip \
-    . /opt/venv/bin/activate && \
```
Comment (hickeyma) [inline docker/Dockerfile:104]: Good catch on duplication from previously.

File: docker/Dockerfile [hunk 6]
```diff
@@ -116,16 +112,15 @@ WORKDIR /workspace/LMCache
 # Stable vLLM release will be used if nightly build is not available.
 RUN --mount=type=cache,target=/root/.cache/ccache,id=ccache \
     --mount=type=cache,target=/root/.cache/uv,id=uv-cache,sharing=locked \
-    . /opt/venv/bin/activate && \
```
Comment (hickeyma) [inline docker/Dockerfile:119]: Good catch on duplication from previously.

File: docker/Dockerfile [hunk 7]
```diff
@@ -116,16 +112,15 @@ WORKDIR /workspace/LMCache
...
     uv pip install --prerelease=allow 'vllm[runai,tensorizer]' \
-        --extra-index-url https://wheels.vllm.ai/nightly \
-        --index-strategy unsafe-best-match && \
+    --extra-index-url https://wheels.vllm.ai/nightly \
+    --index-strategy unsafe-best-match && \
```
Comment (hickeyma) [inline docker/Dockerfile]: Rollback to original indentation

File: docker/Dockerfile [hunk 8]
```diff
@@ -24,8 +24,8 @@ RUN echo 'tzdata tzdata/Areas select America' | debconf-set-selections \
     && echo 'tzdata tzdata/Zones/America select Los_Angeles' | debconf-set-selections \
     && apt-get update -y \
     && apt-get install -y --no-install-recommends \
-        ccache software-properties-common git curl sudo \
-        python3 python3-dev python3-venv python3-pip tzdata \
+    ccache software-properties-common git curl sudo \
+    python3 python3-dev python3-venv python3-pip tzdata \
```
[the remark this answers is not in the document]
Comment (hickeyma) [reply docker/Dockerfile:28]: Leave for now

**Review by hickeyma — CHANGES_REQUESTED** [review]
Thanks @Zerohertz for pushing the PR.

I see the issue you raised #1497. It says that ProductionStack which uses the container image requires the virtualenv to be pathed. Do you mind explaining why it requires it?

File: docker/Dockerfile (hunk 1 above)
Comment (Zerohertz) [reply docker/Dockerfile]: I understand that activating a virtual environment using `. /opt/venv/bin/activate` sets the `VIRTUAL_ENV` environment variable.

**Comment by Zerohertz** [comment]
@hickeyma 

Currently, when using the `lmcache` `vllm-openai` image before our PR is merged, the entrypoint must be set to `/opt/venv/bin/vllm`.

However, the `production-stack` in `vllm` lacks the functionality to change the entrypoint, which is inconvenient.

While we could modify our `production-stack` to add this feature, we believe it's a necessary addition for better compatibility with the official `vllm-openai` image from `vllm`.

> [!NOTE]
> `values.yaml` for `production-stack`
> Screenshot 2025-09-04 at 12 21 30 AM

> [!NOTE]
> Kubernetes manifest of deployments
> Screenshot 2025-09-04 at 12 21 06 AM

File: docker/Dockerfile (hunk 1 above)
Comment (hickeyma) [reply docker/Dockerfile]: From https://docs.python.org/3/library/venv.html:

```
When a virtual environment has been activated, the VIRTUAL_ENV environment variable is set to the path of the environment. Since explicitly activating a virtual environment is not required to use it, VIRTUAL_ENV cannot be relied upon to determine whether a virtual environment is being used.
```

In addition, I don't understand why you are explixitly setting it.

**Comment by hickeyma** [comment]
@Zerohertz Ok, can see the use case for this PR

File: docker/Dockerfile (hunk 1 above)
Comment (Zerohertz) [reply docker/Dockerfile]: That's a valid point. I had defined the path explicitly for clarity, but I agree that it is not a requirement for functionality.
I will proceed with its removal.
Thank you for your insightful review.

**Review by hickeyma — APPROVED** [review]
LGTM, thanks @Zerohertz for pushing PR.

I am going to hold on merging as want to test it locally but have to finish for the day.

**Review by hickeyma — COMMENTED** [review]
Tested locally and ran as expected. Merged. Thanks @Zerohertz
